core/src/main/java/be/orbinson/aem/dictionarytranslator/services/impl/DictionaryServiceImpl.java

```diff
@@ -1,45 +1,91 @@
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceMetadata;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceUtil;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.api.resource.observation.ResourceChange;
+import org.apache.sling.api.resource.observation.ResourceChangeListener;
 import org.apache.sling.jcr.resource.api.JcrResourceConstants;
 import org.jetbrains.annotations.NotNull;
-import org.jetbrains.annotations.Nullable;
 import org.osgi.service.component.annotations.Component;
 import org.osgi.service.component.annotations.Reference;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import javax.jcr.RepositoryException;
-import javax.jcr.Session;
-import javax.jcr.security.AccessControlManager;
-import javax.jcr.security.Privilege;
-import java.util.*;
-import java.util.concurrent.atomic.AtomicReference;
+import com.day.cq.commons.jcr.JcrConstants;
+import com.day.cq.commons.jcr.JcrUtil;
+import com.day.cq.replication.ReplicationActionType;
+import com.day.cq.replication.ReplicationException;
+import com.day.cq.replication.Replicator;
 
-import static be.orbinson.aem.dictionarytranslator.utils.DictionaryConstants.*;
-import static org.apache.jackrabbit.JcrConstants.*;
+import be.orbinson.aem.dictionarytranslator.exception.DictionaryException;
+import be.orbinson.aem.dictionarytranslator.services.DictionaryService;
+import be.orbinson.aem.dictionarytranslator.utils.DictionaryConstants;
 
-@Component
-public class DictionaryServiceImpl implements DictionaryService {
+@Component(property = { ResourceChangeListener.PATHS + "=/",
+ResourceChangeListener.CHANGES + "=ADDED" ,
+ResourceChangeListener.CHANGES + "=REMOVED",
+ResourceChangeListener.CHANGES + "=CHANGED"})
+public class DictionaryServiceImpl implements DictionaryService, ResourceChangeListener {
 
     private static final Logger LOG = LoggerFactory.getLogger(DictionaryServiceImpl.class);
     private static final String SLING_BASENAME = "sling:basename";
+    // same as https://github.com/apache/sling-org-apache-sling-i18n/blob/3f98ebf430e416226500c2975086423edc29dcb3/src/main/java/org/apache/sling/i18n/impl/JcrResourceBundle.java#L69
+    static final String QUERY_LANGUAGE_ROOTS = "//element(*,mix:language)[@jcr:language]";
 
     @Reference
     private Replicator replicator;
 
+    /**
+     * Cache of language dictionaries, the key is an artifical path in the format {@code <dictionaryPath>/<language>}.
+     * The cache for all languages is automatically invalidated when a resource below the dictionary path was changed.
+     */
+    private final Map<String, Map<String, Message>> messagesPerLanguageDictionary = new HashMap<>();
+
     public boolean isEditableDictionary(Resource resource) {
         String path = resource.getPath();
+        if (getType(resource) != DictionaryType.SLING_MESSAGE_ENTRY) {
```

Comment:
This means we will not allow editing of the languages that are message entry based (so a mixed dictionary type). Is this on purpose? Don't know if this is an edge-case to be honest, if you have a lot of mixed dictionaries

Reply:
We can improve later on to allow editing for individual languages, however I condier this an edge case. Usually below one common container it is either one or the other.